src/types-and-traits.rst

```diff
@@ -2249,219 +2272,242 @@ The :t:`type inference` rules for :t:`[statement]s` are as follows:
-   :t:`floating point type`, then ``ET`` is replaced with ``RT``, effectively
-   changing the :t:`type` of all :t:`[expression]s` that previously held ``ET``.
-
-#. :dp:`fls_acd7b3m1qm3a`
-   If ``ET`` denotes an :t:`integer type variable` and ``RT`` denotes an
-   :t:`integer type`, then ``ET`` is replaced with ``RT``, effectively changing
-   the :t:`type` of all :t:`[expression]s` that previously held ``ET``.
-
-#. :dp:`fls_riivz4mlwr4y`
-   Otherwise this is a static error.
+:dp:`fls_J6ydUCCJp1Sn`
+:t:`Type inference` of :t:`[expression]s` may incorporate an :t:`expected type`,
+derived from the context the :t:`expression` appears in. If the :t:`expression`
+is a :t:`coercion site` or a :t:`coercion-propagating expression`, the :t:`type`
+derived via :t:`type inference` may be coerced to the :t:`expected type`. If no
+:t:`type coercion` to the :t:`expected type` is possible, or the :t:`expression`
+is not a :t:`coercion site` or a :t:`coercion-propagating expression`, the
+inferred :t:`expression` :t:`type` is unified with the :t:`expected type`.
+
+:dp:`fls_FSQqHs8T4bUx`
+The :t:`type inference` rules for :t:`[expression]s` are as follows:
+
+* :dp:`fls_0HHC1iOk5dwz`
+  An :t:`if expression` is inferred by inferring its :t:`subject expression`
+  with an :t:`expected type` of :c:`bool`, then inferring its
+  :t:`block expression` with the :t:`expected type` of the :t:`if expression`.
+  Then, if the :t:`if expression` has an :t:`else expression`, apply the
+  inference rules below to it.
+
+* :dp:`fls_QZWTS0Giy3I3`
+  An :t:`if-let expression` is inferred by inferring its
+  :t:`subject let expression` with the :t:`expected type` set to the :t:`type`
+  of its :t:`pattern`, then inferring its :t:`block expression` with the
+  :t:`expected type` of the :t:`if-let expression`. If the
+  :t:`if-let expression` has an :t:`else expression`, apply the inference rules
+  below to it.
+
+* :dp:`fls_KJsIu1lgVZxP`
+  An :t:`else expression` that is part of an :t:`if expression` or
+  :t:`if-let expression` is inferred as follows:
+
+  * :dp:`fls_KRQxdSav1KBA`
+    If the :t:`else expression` has a :t:`block expression`, infer the
+    :t:`block expression` with the :t:`expected type` of the :t:`if expression`
+    or :t:`if-let expression`.
+
+  * :dp:`fls_Mcpwyvz47SoG`
+    If the :t:`else expression` has an :t:`if expression`, infer that nested
+    :t:`if expression` with the :t:`expected type` of the original
+    :t:`if expression`, then :t:`unify` its :t:`type` with the :t:`type` of
+    the original :t:`if expression` or :t:`if-let expression`.
+
+  * :dp:`fls_34AQ9g7xhdUj`
+    Otherwise, the :t:`else expression` has an :t:`if-let expression`. Infer
+    that nested :t:`if-let expression` with the :t:`expected type` of the
+    original :t:`if expression`, then :t:`unify` its :t:`type` with the
+    :t:`type` of the original :t:`if expression` or :t:`if-let expression`.
```

Comment:
```suggestion
    Otherwise, the :t:`else expression` has an :t:`if let expression`. Infer
    that nested :t:`if let expression` with the :t:`expected type` of the
    original :t:`if expression`, then :t:`unify` its :t:`type` with the
    :t:`type` of the original :t:`if expression` or :t:`if let expression`.
```